### DISCUSSION TITLE
How to add an attribute PredefinedType

### DISCUSSION THREAD

**Discussion by DaivaMarcinkeviciute** [IfcOpenShell/IfcOpenShell#3784] [General]
I have an IFC file with some ifcSpaces that do not have a predefined type. I'd like to add predefined type "GFA" and i tried this:

```
elements = ifcFile.by_type("IfcSpace")
or element in elements:
    element.PredefinedType = 'GFA' 
```

Yet i get an error "list index out of range", as the space doesn't have such attribute.
I found a way to add properties with ifcFile.createIfcPropertySingleValue, but nothing for creating attributes, as they are suposed to be fixed. How can i add PredefinedType?

Thanks in advance

**Comment by aothms** [comment]
In the IFC2X3 version of the schema the attribute does not exist. It's a
new addition in IFC4. You can try to use the Schema migration utilities in
IfcOpenShell to try and upgrade your files.
https://standards.buildingsmart.org/IFC/RELEASE/IFC2x3/TC1/HTML/ifcproductextension/lexical/ifcspace.htm

Sent from a mobile device, excuse my brevity. Kind regards, Thomas

Op do 21 sep. 2023 17:30 schreef DaivaMarcinkeviciute <
***@***.***>:

**Reply by DaivaMarcinkeviciute** [reply]
Thanks for the answer! The IFC Schema was indeed the problem. I used this code to upgrade it to IFC4:
https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/recipes/Migrate.py#L46-L49
Afterwards adding PredefinedType worked.
